File: aoc2025/day11/part1.c
#include <assert.h>
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include <string.h>

typedef struct Edge {
    uint16_t dest;
} Edge;

typedef struct Node {
    Edge edges[24];
    uint8_t num_edges;
} Node;

#define MAX_NODE 17578
#define YOU 17576
#define OUT 17577

size_t idx(char name[3]) {
    size_t index = 0;
    if (!strncmp("you", name, 3)) {
        index = YOU;
    } else if (!strncmp("out", name, 3)) {
        index = OUT;
    } else {
        index = ((name[0] - 'a') * 26 * 26) + ((name[1] - 'a') * 26) +
                (name[2] - 'a');
    }

    return index;
}

Node graph[17578] = {0};

size_t dfs(uint16_t n, uint16_t end, size_t *path_count) {
    if (n == end) {
        return 1;
    }

    if (path_count[n] != (size_t)-1) {
        return path_count[n];
    }

    size_t count = 0;

    for (size_t i = 0; i < graph[n].num_edges; i++) {
        count += dfs(graph[n].edges[i].dest, end, path_count);
    }

    path_count[n] = count;
    return count;
}

int main() {
    char buf[1024];
    while (fgets(buf, sizeof(buf), stdin)) {
        char name[3];
        memcpy(name, buf, 3);
        size_t index = idx(name);

        char *ptr = strtok(buf + 5, " \n");
        do {
            memcpy(name, ptr, 3);
            size_t edge_index = idx(name);

            graph[index].edges[graph[index].num_edges++].dest = edge_index;
        } while ((ptr = strtok(NULL, " \n")));
    }

    size_t path_count[MAX_NODE];
    memset(path_count, -1, MAX_NODE * sizeof(size_t));

    size_t count = dfs(YOU, OUT, path_count);

    printf("Total: %zu\n", count);
}

File: aoc2025/day11/part1.zig
const std = @import("std");

const file = @embedFile(".input.txt");
const nodes = 26 * 26 * 26;

const Node = struct {
    edges: [24]u32,
    num_edges: u32,
};

fn idx(name: []const u8) u32 {
    const one: u32 = name[0] - 'a';
    const two: u32 = name[1] - 'a';
    const thr: u32 = name[2] - 'a';
    return (one * 26 * 26) + (two * 26) + thr;
}

fn dfs(graph: []Node, current: u32, end: u32, cache: []usize) usize {
    if (current == end) {
        return 1;
    }

    if (cache[current] != std.math.maxInt(usize)) {
        return cache[current];
    }

    var count: usize = 0;
    for (graph[current].edges[0..graph[current].num_edges]) |n| {
        count += dfs(graph, n, end, cache);
    }

    cache[current] = count;
    return count;
}

pub fn main() !void {
    var graph: [nodes]Node = undefined;

    var file_iter = std.mem.tokenizeScalar(u8, file, '\n');
    while (file_iter.next()) |line| {
        var line_iter = std.mem.tokenizeAny(u8, line, ": ");
        const node_index = idx(line_iter.next().?);

        graph[node_index].num_edges = 0;
        while (line_iter.next()) |edge| {
            const edge_index = idx(edge);
            graph[node_index].edges[graph[node_index].num_edges] = edge_index;
            graph[node_index].num_edges += 1;
        }
    }

    var cache = [_]usize{std.math.maxInt(usize)} ** nodes;
    const answer = dfs(&graph, idx("you"), idx("out"), &cache);
    std.log.debug("Answer {}", .{answer});
}
